x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/CachedBlobContainerIndexInput.java

```diff
@@ -174,29 +174,130 @@ protected void readInternal(final byte[] buffer, final int offset, final int len
+                    return;
+                }
+
+                final long rangeStart = range.v1();
+                final long rangeEnd = range.v2();
+                final long rangeLength = rangeEnd - rangeStart;
+
+                logger.trace(
+                    "prefetchPart: prewarming part [{}] bytes [{}-{}] by fetching bytes [{}-{}] for cache file [{}]",
+                    part,
+                    partRange.v1(),
+                    partRange.v2(),
+                    rangeStart,
+                    rangeEnd,
+                    cacheFileReference
+                );
+
+                final FileChannel fc = cacheFile.getChannel();
+                assert assertFileChannelOpen(fc);
+                final byte[] copyBuffer = new byte[Math.toIntExact(Math.min(COPY_BUFFER_SIZE, rangeLength))];
+
+                long totalBytesRead = 0L;
+                long remaining = rangeEnd - rangeStart;
+                final long startTimeNanos = stats.currentTimeNanos();
+                try (InputStream input = openInputStream(rangeStart, rangeLength)) {
+                    while (remaining > 0L) {
+                        assert totalBytesRead + remaining == rangeLength;
+                        final int bytesRead = readSafe(input, copyBuffer, rangeStart, rangeEnd, remaining, cacheFileReference);
+                        final long readStart = rangeStart + totalBytesRead;
+                        cacheFile.fetchRange(readStart, readStart + bytesRead, (start, end) -> {
+                            logger.trace(
+                                "prefetchPart: range [{}-{}] of file [{}] is now available in cache",
+                                start,
+                                end,
+                                fileInfo.physicalName()
+                            );
+                            return Math.toIntExact(end - start);
+                        }, (start, end) -> {
+                            final ByteBuffer byteBuffer = ByteBuffer.wrap(
+                                copyBuffer,
+                                Math.toIntExact(start - readStart),
+                                Math.toIntExact(end - start)
+                            );
+                            final int writtenBytes = positionalWrite(fc, start, byteBuffer);
+                            logger.trace(
+                                "prefetchPart: writing range [{}-{}] of file [{}], [{}] bytes written",
+                                start,
+                                end,
+                                fileInfo.physicalName(),
+                                writtenBytes
+                            );
+                        });
+                        totalBytesRead += bytesRead;
+                        remaining -= bytesRead;
+                    }
+                    final long endTimeNanos = stats.currentTimeNanos();
+                    stats.addCachedBytesWritten(totalBytesRead, endTimeNanos - startTimeNanos);
```

Comment:
I'm confused by the stat accounting here as I thought that not all `totalBytesRead` have been written to filechannel (when other searches have filled the bytes e.g.)?

Reply:
😕 I thought the intention is for this stat to record the bytes we downloaded from the blob store and so I intended not to account solely for the bytes we actually wrote to the cache file. But I think it's not an accurate count of the bytes downloaded either. I can make it account only for the bytes written by the prewarmer. @tlrx WDYT?

Reply:
I think that `cached_bytes_written` should reflect the number of bytes effectively written in cache. It reflected the number of bytes downloaded (only when the cache range size and the `readBlobPreferredLength` have the same value) but I think we should compute the download stats differently.

Reply:
Ok, I adjusted this in cd793b3 and am checking that the tests are still solid now.